crates/federated-server/src/server/gateway.rs

```diff
@@ -76,20 +82,163 @@ pub(super) async fn generate(
         &config,
         version_id,
         hooks,
-        extensions,
+        Default::default(),
     )
     .await?;
 
     if let Some(trusted_documents) = trusted_documents {
         runtime.trusted_documents = trusted_documents;
     }
 
-    let schema = engine::Schema::build(config, schema_version, &Default::default())
+    let extension_catalog = create_extension_catalog()?;
+
+    let schema = engine::Schema::build(config, schema_version, &extension_catalog)
         .map_err(|err| crate::Error::SchemaValidationError(err.to_string()))?;
 
+    if let Some(extensions) = dbg!(create_wasi_extension_configs(
+        &extension_catalog,
+        gateway_config,
+        &schema
+    )) {
+        runtime.extensions =
+            WasiExtensions::new(access_log, extensions).map_err(|e| Error::InternalError(e.to_string()))?;
+    }
+
     Ok(Engine::new(Arc::new(schema), runtime).await)
 }
 
+fn create_wasi_extension_configs(
+    extension_catalog: &ExtensionCatalog,
+    gateway_config: &Config,
+    schema: &engine::Schema,
+) -> Option<Vec<ExtensionConfig>> {
+    let mut wasi_extensions = Vec::with_capacity(extension_catalog.len());
+
+    for (id, extension) in extension_catalog.iter().enumerate() {
+        let Some(extension_config) = gateway_config.extensions.as_ref().and_then(|h| {
+            let config = h.get(&extension.manifest.name);
+            config.filter(|e| e.version().matches(&extension.manifest.version))
+        }) else {
+            continue;
+        };
+
+        let extension_type = match &extension.manifest.kind {
+            extension_catalog::Kind::FieldResolver(_) => ExtensionType::Resolver,
+        };
+
+        let wasi_config = WasiExtensionsConfig {
+            location: extension.wasm_path.clone(),
+            networking: extension_config.networking(),
+            stdout: extension_config.stdout(),
+            stderr: extension_config.stderr(),
+            environment_variables: extension_config.environment_variables(),
+        };
+
+        let extension_id = ExtensionId::from(id);
+
+        let mut config = ExtensionConfig {
+            id: extension_id,
+            name: extension.manifest.name.clone(),
+            extension_type,
+            schema_directives: Vec::new(),
+            max_pool_size: extension_config.max_pool_size(),
+            wasi_config,
+        };
+
+        for subgraph in schema.subgraphs() {
```

Comment:
Could we do this without `O(n^2)` double loop?

Reply:
I don't see how, if you need to access every directive inside every subgraph. Extension schema directives are nested inside the subgraphs, right?

Reply:
Yeah, maybe it's not an issue. Just bothers me to write this code.

Reply:
It's not `O(n^2)` it's `O(n)`, you're iterating over every extension schema directive once.

Reply:
ah didn't see the outer loop, my bad

Reply:
You could just remove it from this loop and do `gateway_config.extensions.get(catalog[extension_id].manifest.name).schema_directives.push(directive)`

Reply:
Think you should rely on `ExtensionId` defined by the catalog inside the wasi stuff, so just have it in the same order. That way you bypass the hashmap stuff

Reply:
The issue is I do need the name of the subgraph too...